processor/src/main/java/org/mapstruct/ap/model/BeanMappingMethod.java

```diff
@@ -121,17 +126,66 @@ public BeanMappingMethod build() {
                 if ( beanMapping != null && beanMapping.getResultType() != null ) {
                     resultType = ctx.getTypeFactory().getType( beanMapping.getResultType() );
                     if ( !resultType.isAssignableTo( method.getResultType() ) ) {
-                        ctx.getMessager().printMessage( method.getExecutable(),
+                        ctx.getMessager().printMessage(
+                            method.getExecutable(),
                             beanMappingPrism.mirror,
-                            Message.BEANMAPPING_NOT_ASSIGNABLE, resultType, method.getResultType());
+                            Message.BEANMAPPING_NOT_ASSIGNABLE, resultType, method.getResultType()
+                        );
                     }
                 }
             }
 
+            sortPropertyMappingsByDependencies();
+
             return new BeanMappingMethod( method, propertyMappings, factoryMethod, mapNullToDefault, resultType );
         }
 
         /**
+         * Sources the given mappings as per the dependency relationships given via {@code dependsOn()}. If a cycle is
+         * detected, an error is reported.
+         */
+        private void sortPropertyMappingsByDependencies() {
+            final GraphAnalyzer graphAnalyzer = new GraphAnalyzer();
+
+            for ( PropertyMapping propertyMapping : propertyMappings ) {
+                graphAnalyzer.addNode( propertyMapping.getName(), propertyMapping.getDependsOn() );
+            }
+
+            graphAnalyzer.analyze();
+
+            if ( !graphAnalyzer.getCycles().isEmpty() ) {
+                Set<String> cycles = new HashSet<String>( graphAnalyzer.getCycles().size() );
```

Comment:
> Interesting. When will this resizing occur?

After the load factor (capacity : number of items -- default is 75%) was reached.

> I've changed it to use the default constructor, this should be good enough.

Right. And it's for the error-case anyway.